### PR TITLE
Set NodeReady=False when docker is dead

### LINKED ISSUE #6285: kubernetes is not correctly removing nodes in faulty state
in my kubernetes lab cluster (0.13.2) i'm having a issue where faulty nodes are not being changed to not ready status.
I only detected this because docker service wasn't working on that node (which depends on flannel.service [that depends on the network interface where 172.17.8.104 was configured])

```
$ date
Wed Apr  1 11:55:11 WEST 2015

$ kubectl get minion
NAME                LABELS              STATUS
172.17.8.102        <none>              Ready
172.17.8.103        <none>              Ready
172.17.8.104        <none>              Ready

$ kubectl describe minions 172.17.8.104
Name:	172.17.8.104
Conditions:
  Type		Status	LastProbeTime				LastTransitionTime			Reason					Message
  Ready 	Full 	Wed, 01 Apr 2015 10:12:22 +0000 	Mon, 01 Jan 0001 00:00:00 +0000 	kubelet is posting ready status 	
Addresses:	172.17.8.104
Capacity:
 cpu:		1
 memory:	1021908Ki
Pods:		(0 in total)
```

### PR DESCRIPTION
**Pull request by wojtek-t**
Fix #6285 

cc @davidopp @vmarmol

### CODE REVIEW

**Comment by wojtek-t** [comment]
Ref #7222

**Comment by wojtek-t** [comment]
e2e tests are passing:

Ran 38 of 43 Specs in 1474.965 seconds
SUCCESS! -- 38 Passed | 0 Failed | 1 Pending | 4 Skipped I0505 13:31:05.709929   18131 driver.go:96] All tests pass

**Comment by wojtek-t** [comment]
cc @dchen1107 (since @vmarmol seems to be OOO today)

**Comment by vmarmol** [comment]
LGTM

**Comment by vmarmol** [comment]
Only thing that worries me about this approach is whether we have a lot of transient errors due to this. Setting the status won't remove scheduled pods so maybe it isn't much of a concern. We may not be a schedulable node for a bit.

**Comment by dchen1107** [comment]
I have the same concern as @vmarmol above. We observed that Docker might be temporiarily unavailable a lot under some average workload pressure even, we should tolerate such failures at our system. My concern is that we might ping-pong those workloads due to switch node status. here.

**Comment by dchen1107** [comment]
My problem is that we just simply query docker version, it might run into "resource temporarily unavailable" (forget if this is exact err throwed by docker or not) a lot.

**Comment by davidopp** [comment]
NodeController won't evict pods until node has been NodeReady=False for 5 minutes. Are the transient Docker unavailability longer than 5 minutes?

(Though it still may be a problem from a UX perspective. Maybe it's better to mask transient Docker problem on the Kubelet side so user doesn't get confused.)

**Comment by dchen1107** [comment]
@davidopp I agreed with you on this. Kubelet should have some logic inside to decide when docker is not available. Today monit peridically check docker's health, and restart it if it is unresponsive for a certain time (forget the time interval I initially set.) We might should ask kubelet to do such thing instead of monit. But if we take such position, what about other container Runtime? rkt + systemd?

**Comment by wojtek-t** [comment]
Do you have any suggestions how to fix this?

**Comment by vmarmol** [comment]
We should only surface the failure when we've observed Docker to be down
for 5min or so. Keep us from flapping due to transient issues. We may need
to ping Docker at a higher interval than we do for node status.

**Comment by yujuhong** [comment]
Just FYI, there is a function in kubelet.go checking whether docker is alive at startup. You could probably repurpose that for your use. https://github.com/GoogleCloudPlatform/kubernetes/blob/master/pkg/kubelet/kubelet.go#L100

**Comment by wojtek-t** [comment]
@yujuhong thanks - I will try to fix this today.

**Comment by wojtek-t** [comment]
I've update a PR so that the failure will be reported only if Docker didn't respond during last 5 minutes. 

PTAL

**Comment by wojtek-t** [comment]
I'm running e2e tests - will post the results in few minutes.

File: pkg/kubelet/kubelet.go [hunk 1]
```diff
@@ -1444,6 +1455,15 @@ func (kl *Kubelet) GetPodByName(namespace, name string) (*api.Pod, bool) {
 	return kl.podManager.GetPodByName(namespace, name)
 }
 
+func (kl *Kubelet) updateRuntimeUp() {
```
Comment (vmarmol) [inline pkg/kubelet/kubelet.go]: Looks like we don't actually call this anywhere in the Kubelet?

File: pkg/kubelet/kubelet.go [hunk 2]
```diff
@@ -1444,6 +1455,15 @@ func (kl *Kubelet) GetPodByName(namespace, name string) (*api.Pod, bool) {
...
+	err := waitUntilRuntimeIsUp(kl.containerRuntime, 0)
```
Comment (vmarmol) [inline pkg/kubelet/kubelet.go]: I don't think we can use 0 as a value here. I think [the check](https://github.com/wojtek-t/kubernetes/blob/kubelet_removing_nodes/pkg/kubelet/kubelet.go#L103) will always fail (or be racy)

File: pkg/kubelet/kubelet.go (hunk 1 above)
Comment (wojtek-t) [inline pkg/kubelet/kubelet.go]: Good catch! Thanks.

File: pkg/kubelet/kubelet.go (hunk 2 above)
Comment (wojtek-t) [inline pkg/kubelet/kubelet.go]: Done.

**Comment by wojtek-t** [comment]
I will rerun e2e tests and will update this thread.

**Comment by wojtek-t** [comment]
It seems that e2e tests are failing. I will take a deeper look tomorrow, but if you have any ideas what can be wrong please let me know.

**Comment by wojtek-t** [comment]
Ran 38 of 43 Specs in 798.277 seconds
FAIL! -- 22 Passed | 16 Failed | 1 Pending | 4 Skipped F0506 19:35:42.270701   31759 driver.go:94] At least one test failed


The first 22 tests passed, the next 16 failed :(

File: pkg/kubelet/kubelet.go [hunk 3]
```diff
@@ -1444,6 +1456,15 @@ func (kl *Kubelet) GetPodByName(namespace, name string) (*api.Pod, bool) {
 	return kl.podManager.GetPodByName(namespace, name)
 }
 
+func (kl *Kubelet) updateRuntimeUp() {
+	err := waitUntilRuntimeIsUp(kl.containerRuntime, 25*time.Millisecond)
```
Comment (vmarmol) [inline pkg/kubelet/kubelet.go]: We can probably wait longer since we have a thread for this.

Comment (wojtek-t) [inline pkg/kubelet/kubelet.go]: Done.

**Comment by wojtek-t** [comment]
I've run e2e tests - all tests passed:

Ran 38 of 43 Specs in 1593.325 seconds
SUCCESS! -- 38 Passed | 0 Failed | 1 Pending | 4 Skipped I0507 10:56:37.327217    5607 driver.go:96] All tests pass

**Comment by vmarmol** [comment]
LGTM, thanks @wojtek-t!

**Pull request merged**